### DISCUSSION TITLE
Multi-thread Query Execution (Morsel-driven Parallelism)

### DISCUSSION THREAD

**Discussion by csruiliu** [duckdb/duckdb#6632] [Q&A] [answered]
I was exploring the source code and want to figure out how the multi-thread query execution (or morsel-driven alike parallelism) has been implemented; I would like to collect some hints/insights here; thanks a lot for your help and please accept my apologies if I asked a dull question. 

I ran the following query using Python APIs with multi-thread configuration as an example (let me know if I did it wrong or chose a poor example). The lineitem is with SF10. I expected there would be 4 threads to read the file simultaneously and combine the results at the sink of the pipeline (my rig has 10900K and 64GB memory).

```
query = "SELECT  avg(L_DISCOUNT) as AVG_DISC  FROM  'TPCH_DATAPATH/lineitem.parquet'"
db_conn = duckdb.connect(database=':memory:')
db_conn.execute("PRAGMA threads=4")
results = db_conn.execute(query).fetchdf()
```

1. The task execution is essentially triggered in `CompletePendingQuery` in `pyconnection.cpp` after query parsing and binding. There is a loop to keep calling ExecuteTask() until the results are ready. 

2. By tracking the source code, it seems that the `ScheduleParallel` will be called when a `PipeInitializedEvent` is first created and scheduled in a MoodyCamel-based concurrent queue for a pipeline.  

3. `ScheduleParallel` can generate a number of PipelineTasks and put them into the queue for reading the parquet file. The number of PipelineTask is determined by `max_thread` (which always gives me 1, no matter how much I set, I assume we could have more reading threads to enable parallel reading). 

4. The loop in `CompletePendingQuery` will keep calling the PipelineTasks one by one for execution (`PipelineExecutor::Execute`). The PipelineTask will read the data source (`PipelineExecutor::FetchFromSource`) and aggregate them (`PipelineExecutor::PushFinalize`) for the final result.

If so, the tasks in the queue seem to be executed sequentially by being called through the loop rather than in parallel. I assume I misunderstood something here, and I would like to get some feedback or insights about how parallel query execution is implemented in DuckDB. Thanks!!

**Comment by lokax** [comment]
Multiple threads pull tasks from the queue and execute them. Any pipeline task with a breaker will enter the thread pool for execution. So AGG::Sink is parallel, and the max thread of AGG::Source is determined by the amount of aggregated data(the max thread of AGG::Sink is determined by num_row_group of parquet. ). Since AGG::Source's pipeline has no pipeline breaker, it can only be executed continuously in the main thread by sequential. You can try materialize query result,.it will add PhysicalCollector as pipeline breaker.

**Reply by csruiliu** [reply]
Thanks a lot for your answers! I have some following questions:

1. How do the multiple threads pull tasks from the queue? if, as I mentioned before, the task execution is triggered via a loop in `CompletePendingQuery` in `pyconnection.cpp`, it seems that there is only one main thread for execution.
2. Could you please be more specific about the `breaker`? 
3. If the `breakers` determine the execution parallelism, can I understand that the current DuckDB doesn't support reading in parallel for the simple query I provided even if the `lineitem.parquet` could be extremely large? 

Thanks!

**Reply by Mytherin** [reply]
Parquet files are parallelized over row groups - your Parquet file likely has a single giant row group which is preventing parallel execution. Have a look at the metadata functions to figure out how to inspect that: https://duckdb.org/docs/data/parquet/metadata

**Reply by csruiliu** [reply]
> Parquet files are parallelized over row groups - your Parquet file likely has a single giant row group which is preventing parallel execution. Have a look at the metadata functions to figure out how to inspect that: https://duckdb.org/docs/data/parquet/metadata

Thanks for your response! Assuming we will have multiple row groups, then we should have multiple PipelineTasks in the queue for execution, I assume DuckDB will have a thread pool (or -alike) mechanism to allocate multiple threads and run these tasks in parallel. However,  It seems like the loop in `CompletePendingQuery` in `pyconnection.cpp` will trigger it (if we use Python APIs), specifically, in each loop, `Executor::ExecuteTask` in `executor.cpp` will be called, and a new task will be pulled from the queue for execution until return `RESULT_READY`. Everything looks like sequential execution to me, and if we have multiple threads, how do they sync up with each to get the final results accurately and efficiently? 

I tried to search for answers (including but not limited to the official documentation and your presentation on youtube), but no clear clue, and I didn't trust what ChatGPT said about DuckDB :). I would appreciate it if you could provide some insights or documents so that I can figure out parallel execution better.

**Reply by Mytherin** [reply]
There are two paths to executing tasks. The one you mention is for executing tasks in the main thread - and is indeed sequential. There is also a thread-pool of threads that continuously run and execute tasks in the background through `TaskScheduler::ExecuteForever`.

**Reply by csruiliu** [reply]
Thanks, I will start from there. Marked as answered but also expect more insightful communication with you and the community.

**Comment by csruiliu** [comment]
Thanks again for your response. I did enable the parallel processing in a way for the aforementioned simple queue by splitting the parquet file into multiple row groups (row group size 100000). After revisiting some relevant source codes, I have two following questions:

1. I was wondering how the `ExecuteForever` function has been called during runtime. it seems that it is only involved by the chain of `DatabaseInstance::Initialize` -> `SetThreads` -> `SetThreadsInternal` -> `make_unique<thread>(ThreadExecuteTasks, ..., ...)`, but `ExecuteForever` is indeed invoked during each task polling in the loop, namely, 1 thread executes a task in PROCESS_PARTIAL mode, all other `n-1` threads execute in PROCESS_ALL mode.

2. I was also curious why the `WorkOnTasks()` function had been called in `CancelTasks`; to finish any tasks left (if so, why don't reset or even destroy them) or some other underlying purposes?

Thanks in advance for any comments!

**Comment by lokax** [comment, marked as answer]
1. ExecuteForever continuously loops to pull tasks from the queue until the thread is canceled
2. One query one executor. Queue is shared by multiple executor (I'm not really sure). So the queue cannot be simply reseted, the main thread must pull the tasks belonging to the current executor from the queue (marked by producer token). If the canceled task cannot be removed from the queue in time, the threads in the thread pool will still execute the canceled task

**Reply by csruiliu** [reply]
Thanks @lokax for your response. No expert here, but I was actually wondering how `ExecuteForever` is invoked to continuously loops to pull the tasks. As I mentioned before, my understanding, which I knew was wrong though, is the `ExecuteForever` is only called at the chain starting from `DatabaseInstance::Initialize` and put a couple for thread in the `vector<unique_ptr<SchedulerThread>> threads` in `task_scheduler.hpp`, then how the `threads` or `ExecuteForever` handle the tasks during runtime. I failed to find the pieces of the code. Could you provide some hints? 

Also, I assume each pipeline will have a dedicated buffer pool, and the metadata of each pipeline can be stored so that DuckDB will know where to find the intermediate results to merge/combine for a common sink. Is my understanding correct?

I assume there is no such buffer pool for each thread, if so, I was also wondering then how the results from different threads be merged efficiently.  

I mentioned @Mytherin as well for any further comments. Really appreciate it.

**Reply by lokax** [reply]
1. ExecuteForever will only be called once when setting the number of threads. The function will continue to pull the task from the queue and do execution. ExecuteForever is the entry function of the thread, no magic here.
2. I think buffer pool is global. Sometimes OperatorState will store BlockHandle from buffer pool . So  you can know where the data is. Merge/Combine sometimes just copy the pointer (8 bytes) of the local memory block into the GlobalOperatorsState. So it is efficient, because the overhead of the large memory copy is avoided. The last thread that completes the task will call Finalize() to build some global data structure. But different operators have different implementation, the above is just a simple description.

**Reply by csruiliu** [reply]
@lokax Thanks for your prompt response. 

1. Yeah, I understand `ExecuteForever` will only be called once and put all threads to `vector<unique_ptr<SchedulerThread>> threads` for future use, but could you please be more specific about "ExecuteForever is the entry function of the thread" and show me some example code for better understanding?

2. So, each pipeline has its own buffer pool, and if a single pipeline can support multi-thread execution, the threads actually share the buffer pool of the pipeline (I believe there will be some concurrency control), is it right?

**Reply by lokax** [reply]
1. You can search `TaskScheduler::ExecuteForever`
2. One `DatabaseInstance` one `BufferManager`

**Reply by csruiliu** [reply]
1. Yeah, that is the part I don't fully understand. In the function, each thread in the `threads` vector will process in `PROCESS_ALL` mode in the background, but it seems that there is a `main` thread process in the front but in `PROCESS_PARTIAL` (i.e., chunk by chunk), why is that?

2. I will take a further look.

**Reply by lokax** [reply]
@csruiliu (PARTIAL_MODE). This is a demand-driven situation, you can only pull a batch of data for processing instead of all the data. StreamQueryResult, the root pipeline can only be processed by the main thread. Because the main thread is user-controllable, users can call ExecuteTask once to calculate a batch of data when they need data. MaterialQueryResult, it will add an operator as a pipeline breaker. the root pipeline will be processed by the main thread together with threads from the thread pool. 

For example, two tables do hash join. There will be two pipelines.
1. table scan -> hash build
2. table scan -> hash probe -> projection

1 has pipeline breaker.
2 is the root pipeline. 
When you only want a batch of data, 2 cannot be processed by threads in the thread pool, because their mode is ALL_MODE

**Reply by lokax** [reply]
https://youtu.be/MA0OsvYFGrc

**Reply by csruiliu** [reply]
Thanks @lokax, I really appreciate your help! I think I understand the situation of multiple pipelines, but when I run a query with only one pipeline using 4 threads, like the following one (row group size in parquet file is 100000) 

```
SELECT  avg(L_DISCOUNT) as AVG_DISC,
               sum(L_QUANTITY) as SUM_QTY
FROM    'TPCH_DATAPATH/lineitem.parquet
```

According to the printout, I still have 1 main thread, which processes the data chunk by chunk (PROCESS_PARTIAL mode), and 3 threads running by ExecuteForever in PROCESS_ALL mode. I expect each thread may handle a number of row groups when we have multiple threads and multiple row groups, so I was wondering why the other 3 threads will process in PROCESS_ALL mode.

**Reply by lokax** [reply]
PROCESS_ALL just means that the thread A needs to perform the task until all the data is processed, but not all the data is processed by the thread A. POROCESS_PARTIAL means the thread A needs to perform the task until PARTIAL_CHUNK_COUNT(50 * 2048) is processed. you can look pipeline.cpp and pipeline_executor.cpp.

**Reply by csruiliu** [reply]
Thanks for all your responses, really appreciate it! I will revisit more source codes for a better understanding.
